A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+it will be sent regardless of whether the client actually needs it.  
+This could conceivably lead to wasted bandwidth, CPU, and memory if the backend is sending data 
+that the client does not need, but this case should be fairly rare; 
+backends will generally not be written to populate data that they do not need.  
+Also, we want to avoid the complexity of introducing some kind of negotiation for the client to 
+tell the backend that it wants this data.
+
+#### The Client API
+LB policies will have the ability to access the load report data 
+as part of intercepting the trailing metadata sent from the server.
+Note that the API must be structured such that if multiple LB policies are interested in the data (e.g., both xDS and WRR),
+the protobuf deserialization needs to happen only once. The exact API for this may differ across languages, 
+but here are a couple of possible approaches:
+* The deserialized data can be attached to the metadata element, so that it can be deserialized once and accessed multiple times.
+* The deserialized data can be stored in the call data and accessed by the LB policy code when it intercepts the trailing metadata.
+
+In addition, we also discussed the possibility of having the LB policy get a separate callback for the load report data, 
+distinct from the callback it gets to intercept trailing metadata.  
+However, we think this is not a good idea, since it is likely to cause ordering problems
+(there is no guarantee of the order in which the callbacks would be seen by the LB policy) 
+and may incur additional synchronization overhead.
+
+#### The Server API 
+Using a server interceptor, it is convenient to attach the metrics as part of the
+trailing metadata of the server call and send back with the RPC response.
+
+Define a metric recorder that is unique per RPC. Server applications should obtain a reference and 
+use it to record metrics specifically for the current RPC call. Implementing this may leverage a 
+[context](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html) or similar mechanism, 
+and a language should implement one if not already existed.
+The metric recoder looks like this:
+
+```Java
+public final class CallMetricRecorder {
+  // Records a request cost metric measurement for the call.
+  public CallMetricRecorder recordRequestCostMetric(String name, double value);
+
+  // Records a utilization metric measurement for the call.
+  public CallMetricRecorder recordUtilizationMetric(String name, double value);
+
+  //Records the CPU utilization metric measurement for the call. 
+  public CallMetricRecorder recordCPUUtilizationMetric(double value);
+
+  // Records the CPU utilization metric measurement for the call. 
+  public CallMetricRecorder recordMemoryUtilizationMetric(double value);
+
+  // Returns the call metric recorder attached to the current call.
+  public static CallMetricRecorder getCurrent();
+}
+```
+
+All the methods should be thread safe.
```

Comment:
Thread-safety is an implementation detail; it does not need to be mentioned here (but it can be mentioned in the Java-specific appendix).